### PR TITLE
Add notifications when Target drops LogEvent or cache size grows, grow *2 instead of +1 

### PR DESCRIPTION
**Pull request by Pomoinytskyi**
Gives ability to be notified that some LogEvents was not stored, and AsyncRequestQueue growing.

### CODE REVIEW

**Comment by snakefoot** [comment]
Bad idea to start allocating EventArgs when the system is already stressed, and no one actually listens to the events.

**Comment by Pomoinytskyi** [comment]
@snakefoot 
I am agree that it is too late to handle situation is we are loosing LogEvents. We have system which can have very high logging acceleration, and we need to know that we are loosing messages.
Presumably good to notify when queue full up to some percentage.

**Comment by snakefoot** [comment]
@Pomoinytskyi Was also thinking you could check if anyone actually listens for the event-listener, before start allocating objects.

**Comment by snakefoot** [comment]
@Pomoinytskyi Could be nice that the "Grow"-event only occurred when it actually grew, and not when reaching the initial requestLimit (Maybe update based on queue-capacity)

**Comment by Pomoinytskyi** [comment]
@snakefoot 
Queue has only Count property, it doesn't provide actual size. I can calculate it based on internal queue algorithms but it will be approximated value. 
Or raise event when queue Count became bigger than maximum Count during working time. 
What do you think?

**Comment by snakefoot** [comment]
I would probably multiply RequestLimit with 2 when reaching the limit


Sent from my Samsung device


-------- Original message --------

File: src/NLog/Targets/Wrappers/AsyncRequestQueue-T.cs [hunk 1]
```diff
@@ -185,5 +197,38 @@ public void Clear()
...
+        private void OnLogEventDropped(AsyncLogEventInfo logEventInfo)
+        {
+            var logEventDropped = LogEventDropped;
+            if (logEventDropped == null)
+            {
+                return;
+            }
+
+            LogEventDroppedEventArgs eventArg = new LogEventDroppedEventArgs(logEventInfo);
+            logEventDropped.Invoke(this, eventArg);
```
Comment (304NotModified) [inline src/NLog/Targets/Wrappers/AsyncRequestQueue-T.cs]: I prefer ` LogEventDropped?.Invoke(this, new LogEventDroppedEventArgs(logEventInfo));`

the same for `OnLogEventQueueGrows` 

less is more :)

**Comment by 304NotModified** [comment]
> I would probably multiply RequestLimit with 2 when reaching the limit

this isn't a change that is for the notification, isn't? I think this should be a different PR

**Comment by snakefoot** [comment]
> this isn't a change that is for the notification, isn't? I think this should be a different PR?

Seems silly to make growth notification when not growing. Ex. RequestLimit set to 10000, when count 10001 is reached then notification is correctly made. Then one is processed and then one is inserted, thus again count 10001 is reached causing the same notification, but no growth has occurred.

**Comment by Pomoinytskyi** [comment]
> Then one is processed and then one is inserted, thus again count 10001 is reached causing the same notification, but no growth has actually occurred.

I multiply RequestLimit if count equals to RequestLimit and not decrease if items was processed. 
On the next iteration we compare _logEventInfoQueue.Count with new RequestLimit and it should not raise event while _logEventInfoQueue.Count < RequestLimit (already doubled)

```
if (_logEventInfoQueue.Count >= RequestLimit)
      {
              InternalLogger.Debug("Async queue is full");
              switch (OnOverflow)
              {
...
                      case AsyncTargetWrapperOverflowAction.Grow:
                            InternalLogger.Debug("The overflow action is Grow, adding element anyway");
                            RequestLimit *= 2;
                            OnLogEventQueueGrows(RequestCount + 1);
                            break;
...
```
Correct me please, maybe i miss something.

**Comment by Pomoinytskyi** [comment]
> Then one is processed and then one is inserted, thus again count 10001 is reached causing the same notification, but no growth has actually occurred.

I multiply RequestLimit if count equals to RequestLimit and not decrease if items was processed. 
On the next iteration we compare _logEventInfoQueue.Count with new RequestLimit and it should not raise event while _logEventInfoQueue.Count < RequestLimit (already doubled)

```
if (_logEventInfoQueue.Count >= RequestLimit)
      {
              InternalLogger.Debug("Async queue is full");
              switch (OnOverflow)
              {
...
                      case AsyncTargetWrapperOverflowAction.Grow:
                            InternalLogger.Debug("The overflow action is Grow, adding element anyway");
                            RequestLimit *= 2;
                            OnLogEventQueueGrows(RequestCount + 1);
                            break;
...
```
Correct me please, maybe i miss something.

**Comment by snakefoot** [comment]
@Pomoinytskyi Yes that was my idea.

Could you also remove the LogEventDropped-event from Target.cs, and just have it on the AsyncWrapper ?

**Comment by Pomoinytskyi** [comment]
@snakefoot 
I thought that LogEventDropped  event could be reused in other targets,  depends on internal implementation. For example some network target can lost events on network error etc.

**Comment by snakefoot** [comment]
@Pomoinytskyi I was hoping one day, that flow-control for NLog-targets was something done by a wrapper, and not something all targets had to implement themselves. Most flow-control and batching-logic I have seen on NLog-targets are very primitive and doesn't work well with recent inventions like the Internet.

**Comment by snakefoot** [comment]
> Targets/TargetDropEventEventArgs.cs(39,34): error CS1574: XML comment has cref attribute 'LogEventDropped' that could not be resolved

**Comment by Pomoinytskyi** [comment]
@snakefoot 
Sorry, stupid mistake.  Fixed.

**Comment by 304NotModified** [comment]
@snakefoot I think this one is ready to merge, do you agree?

**Comment by snakefoot** [comment]
I think that these events on AsyncTargetWrapper:

public event EventHandler<TargetDropEventEventArgs> LogEventDropped;
public event EventHandler<LogEventQueueGrowEventArgs> EventQueueGrow;

Should forward directly to the AsyncRequestQueue-T event. This will prevent unnecessary EventArgs-allocation.

See also AppDomainWrapper.cs for an example of how this is done:

https://github.com/NLog/NLog/blob/19de4feb0a6f1292e9d0c69da5c75b73efda3ea1/src/NLog/Internal/Fakeables/AppDomainWrapper.cs#L180

P.S. @304NotModified  Curious why you keep piling up new stuff on NLog 4.5, thought it was stabilization period.

**Comment by 304NotModified** [comment]
> P.S. @304NotModified Curious why you keep piling up new stuff on NLog 4.5, thought it was stabilization period.

I don't like open pull requests. Also if I have the time, I prefer to spent it useful.

**Comment by snakefoot** [comment]
> it will be tech debt for reviewing and demotivating for contributors IMO

People waiting for NLog 4.5 RTM within "in weeks", might also be demotivated by hearing the same words 3 months after :)

**Comment by Pomoinytskyi** [comment]
@snakefoot 

> public event EventHandler LogEventDropped;
>public event EventHandler EventQueueGrow;

>Should forward directly to the AsyncRequestQueue-T event. This will prevent unnecessary EventArgs-allocation.

I'd like to have different event, in order to replace sender. I'd like to have AsyncTargetWrapper in sender instead of AsyncRequestQueue-T
What do you think?

**Comment by snakefoot** [comment]
> I'd like to have different event, in order to replace sender. I'd like to have AsyncTargetWrapper in sender instead of AsyncRequestQueue-T

Then AsyncTargetWrapper should only subscribe to AsyncRequestQueue-T, when someone subscribes to AsyncTargetWrapper.

**Comment by snakefoot** [comment]
@Pomoinytskyi Something like this:

```
        public event EventHandler<TargetDropEventEventArgs> LogEventDropped
        {
            add
            {
                if (_logEventDropped == null)
                    RequestQueue.LogEventDropped += OnRequestQueueDropItem;
                _logEventDropped+= value;
            }
            remove
            {
                _logEventDropped -= value;
                if (_logEventDropped== null)
                    RequestQueue.LogEventDropped -= OnRequestQueueDropItem;
            }
        } 
        private event EventHandler<TargetDropEventEventArgs> _logEventDropped;

        private void OnRequestQueueDropItem(object sender, LogEventDroppedEventArgs logEventDroppedEventArgs) 
        {
            _logEventDropped?.Invoke(this, logEventDroppedEventArgs);  // No need to allocate twice
        }
```

**Comment by 304NotModified** [comment]
moved to 4.6 as I'm releasing 4.5 rc7 now.

**Comment by Pomoinytskyi** [comment]
@snakefoot 
Thanks. I am in update process.
I just wondering about construction
` if (_logEventDropped == null)
                    RequestQueue.LogEventDropped += OnRequestQueueDropItem;
                _logEventDropped+= value;`
It is not thread safe, when two threads want to subscribe. But i think it has very low probability.

**Comment by snakefoot** [comment]
> It is not thread safe, when two threads want to subscribe. But i think it has very low probability.

You are very welcome to improve on this. But yes in the case where on is subscribing, and another is unsubscribing then there will be a race-condition. If two threads are subscribing, then no harm is done by subscribing twice, only one subscription will be added (Handled by `EventHandler`-delegate)

**Comment by Pomoinytskyi** [comment]
@snakefoot I think we can neglect this subscribe/unsubscribe race condition. Is it Ok?

**Comment by snakefoot** [comment]
> I think we can neglect this subscribe/unsubscribe race condition. Is it Ok?

Yes fine by me :)

**Comment by snakefoot** [comment]
@Pomoinytskyi Think the automatic unsubscription in `CloseTarget` should be removed.

**Comment by snakefoot** [comment]
@Pomoinytskyi For consistent naming of private variables with underscore prefix, then it should be:

```
private event EventHandler<LogEventDroppedEventArgs> _logEventDroppedEvent;
private event EventHandler<LogEventQueueGrowEventArgs> _eventQueueGrowEvent;
```

Guess we need some stylecop logic :)

**Comment by 304NotModified** [comment]
@Pomoinytskyi

would it be possible to add some unit tests?

**Comment by 304NotModified** [comment]
> @Pomoinytskyi
> 
> would it be possible to add some unit tests?

@Pomoinytskyi polite bump

**Comment by Pomoinytskyi** [comment]
@304NotModified Yes, i will cover it with UT. Hope to the Monday.

**Comment by 304NotModified** [comment]
Thanks in advance!

**Comment by 304NotModified** [comment]
@Pomoinytskyi Any idea when you have time to add some unit tests? Thanks!

**Comment by 304NotModified** [comment]
Hi @Pomoinytskyi ,

We're trying to complete the PRs for NLog 4.6. If you have time, please add some unit tests. Otherwise this will be NLog 4.7

**Comment by Pomoinytskyi** [comment]
Sorry, i was on vacation, today will spend for unit tests.

**Comment by 304NotModified** [comment]
Thanks!

It seems the build is failing now. We need to check that.

**Comment by 304NotModified** [comment]
I see that the appveyor build was stuck (and cancelled.) restarted it

File: src/NLog/Targets/Wrappers/AsyncRequestQueueBase.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+using System;
+
+namespace NLog.Targets.Wrappers
+{
+    using System.Collections.Generic;
+    using NLog.Common;
+
+    internal abstract class AsyncRequestQueueBase
+    {
+        public virtual bool IsEmpty { get; }
```
Comment (snakefoot) [inline src/NLog/Targets/Wrappers/AsyncRequestQueueBase.cs]: Any reason for not making it abstract ?

Comment (Pomoinytskyi) [reply src/NLog/Targets/Wrappers/AsyncRequestQueueBase.cs]: Only reasin is avoid duplicates.

Comment (snakefoot) [reply src/NLog/Targets/Wrappers/AsyncRequestQueueBase.cs]: Not sure I understand "avoid duplicates"

Comment (Pomoinytskyi) [reply src/NLog/Targets/Wrappers/AsyncRequestQueueBase.cs]: I misunderstand  your question.
It should be abstract.

**Comment by 304NotModified** [comment]
@snakefoot I think this is ready for merge, do you agree?

**Comment by 304NotModified** [comment]
ok thanks!

**Comment by 304NotModified** [comment]
FYI, NLog 4.6 is online which includes this feature!

**Pull request merged**